### PR TITLE
Add support for Marketplace Metering Endpoints

### PR DESCRIPTION
**Pull request by carlisom**
The only [documented](http://docs.aws.amazon.com/marketplacemetering/latest/APIReference/API_MeterUsage.html) API found on their public site is `MeterUsage`.

`BatchMeterUsage` and `ResolveCustomer` can be found in the [JavaScript SDK](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/MarketplaceMetering.html).

The base16 library addition is to fix a problem cause by AWS sending non padded base16 encoded customer tokens that are used when calling `ResolveCustomer`.

### CODE REVIEW

File: src/erlcloud_mms.erl [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+%%%------------------------------------------------------------------------------
+
+-type json_term() :: jsx:json_term().
+
+%%%------------------------------------------------------------------------------
+%%% Batch Meter Usage
+%%%------------------------------------------------------------------------------
+
+-spec batch_meter_usage(ProductCode :: binary(),
+                        UsageRecords :: proplist()) -> proplist().
```
Comment (motobob) [inline src/erlcloud_mms.erl]: request uses `json_return()` so spec s are broken here.

File: src/erlcloud_mms.erl [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+default_config() -> erlcloud_aws:default_config().
+
+%%%------------------------------------------------------------------------------
+%%% Shared Types
+%%%------------------------------------------------------------------------------
+
+-type json_term() :: jsx:json_term().
+
+%%%------------------------------------------------------------------------------
+%%% Batch Meter Usage
```
Comment (motobob) [inline src/erlcloud_mms.erl]: could you please add links to AWS API documentation?

File: src/erlcloud_mms.erl [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+%%%------------------------------------------------------------------------------
+%%% Request
+%%%------------------------------------------------------------------------------
+
+-type operation() :: string().
+-type json_return() :: {ok, json_term()} | {error, term()}.
+
+-spec request(aws_config(), operation(), json_term()) -> json_return().
+request(Config, Operation, Json) ->
```
Comment (motobob) [inline src/erlcloud_mms.erl]: what is the point of this wrapper given nobody uses `request2` directly?
please simplify

File: src/erlcloud_mms.erl [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+                               httpc_error_reason = undefined,
+                               response_status = Status,
+                               response_status_line = StatusLine,
+                               response_body = Body}) ->
+    {error, {http_error, Status, StatusLine, Body}};
+request_to_return(#aws_request{response_type = error,
+                               error_type = aws,
+                               httpc_error_reason = Reason}) ->
+    {error, Reason};
+request_to_return(#aws_request{response_type = error,
```
Comment (motobob) [inline src/erlcloud_mms.erl]: how is this clause differ from previous?

File: src/erlcloud_mms.erl [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+        Status >= 400 andalso Status < 500 ->
+            client_error(Request2);
+        Status >= 500 ->
+            Request2#aws_request{should_retry = true};
+        Status < 400 ->
+            Request2#aws_request{should_retry = false}
+    end.
+
+-spec client_error(#aws_request{}) -> #aws_request{}.
+client_error(#aws_request{response_body = Body} = Request) ->
```
Comment (motobob) [inline src/erlcloud_mms.erl]: i really do not like the idea of copy pasting ddb& ddb_streams code here. 
let's generalise it and use one copy

**Comment by carlisom** [comment]
@motobob please have another look

**Pull request merged**